std(docs): clarify how std::fs::set_permisions works with symlinks

lolbinarycat opened the pull request:
fixes https://github.com/rust-lang/rust/issues/75942
fixes https://github.com/rust-lang/rust/issues/124201

<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r? <reviewer name>
-->
<!-- homu-ignore:end -->

lolbinarycat committed:
std(docs): clarify how std::fs::set_permisions works with symlinks

fixes https://github.com/rust-lang/rust/issues/75942
fixes https://github.com/rust-lang/rust/issues/124201

the8472 commented on library/std/src/fs.rs:
```diff
@@ -2980,6 +2980,19 @@ pub fn read_dir<P: AsRef<Path>>(path: P) -> io::Result<ReadDir> {
 ///
 /// [changes]: io#platform-specific-behavior
 ///
+/// # Symlinks
```
This should be nested under platform-specific behavior.

the8472 commented on library/std/src/fs.rs:
```diff
@@ -2980,6 +2980,19 @@ pub fn read_dir<P: AsRef<Path>>(path: P) -> io::Result<ReadDir> {
 /// [changes]: io#platform-specific-behavior
 ///
+/// # Symlinks
+/// On UNIX systems, it is impossible to manipulate the permission bits of a symlink itself[^1].
```
The language is too strong, https://superuser.com/a/1188707

And it's leading with the wrong thing, we should first say that it follows symlinks and what implications that has. The why would be more of a footnote.

the8472 commented on library/std/src/fs.rs:
```diff
@@ -2980,6 +2980,19 @@ pub fn read_dir<P: AsRef<Path>>(path: P) -> io::Result<ReadDir> {
+/// Because of this, on those systems, this function will update the permission bits
+/// of the file pointed to by the symlink.
+///
+/// Note that this behavior can lead to privalage escalation vulnerabilites,
+/// where the ability to write a symlink in one directory allows you to
+/// cause the permissions of another directory to be modified.
```
This is too vague, especially "write a symlink" (create?) and I don't think it's not specific to the target of the symlink being a directory.

lolbinarycat committed:
clarify std::fs::set_permissions symlink behavior

nest under platform-specific behavior,
factor rationale into its own section,
and tweak language.

lolbinarycat commented:
@rustbot review

thomcc commented:
@bors r+ rollup

bors commented:
:pushpin: Commit 37c4a37ca2d94047b984aa16a95800b48255d327 has been approved by `thomcc`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=thomcc 37c4a37ca2d94047b984aa16a95800b48255d327 -->
<!-- homu: {"type":"Approved","sha":"37c4a37ca2d94047b984aa16a95800b48255d327","approver":"thomcc","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
